### PR TITLE
add ucb and ei

### PR DESCRIPTION
**Pull request by nikita-0209**
There was some stuff in the previous code that did not make complete sense to me so I didn't implement them here. My understanding might be totally wrong so I am enlisting them below so that you guys can correct me:
1. The need to scale up the mean and variance in ucb [[link to code line](https://github.com/alexhernandezgarcia/ActiveLearningPipeline/blob/2ae63854ecb6a46a750b1d76755c61c57dc3f2e1/previous_code/models.py#L263)]
2. I wasn't too sure why we should return the uncertainty along with the score [[link](https://github.com/alexhernandezgarcia/ActiveLearningPipeline/blob/2ae63854ecb6a46a750b1d76755c61c57dc3f2e1/previous_code/models.py#L267)] -- like what do we aim to learn from the uncertainty?
3. Taking the negative value of `u` [[link](https://github.com/alexhernandezgarcia/ActiveLearningPipeline/blob/2ae63854ecb6a46a750b1d76755c61c57dc3f2e1/previous_code/models.py#L278)]
4. Also I am not too sure why `l2r()` is applied on `outputs` instead of on `ei` in the previous code of expected improvement [[link](https://github.com/alexhernandezgarcia/ActiveLearningPipeline/blob/main/models.py#L275)]. For now, I've applied it on `ei` as done in the previous cod.

For now, I added the default values in the argument parser [here](https://github.com/alexhernandezgarcia/ActiveLearningPipeline/blob/main/main.py#L232) to our config_test.yaml file. We could modify them depending on our experiments.

For future, it might be a good idea to have a `forward_with_uncertainty()` which could execute the dropout logic instead of writing the `hstack` code repeatedly. (Might be also useful for other acquisition functions like MES.

There will be a merge conflict once the previous PR is merged (change different interfaces of the model into a single kind of interface) -- I will resolve that.

Minor change in proxy.py to bring the inputs on the device as was running into an error when I shifted to `cuda`

### CODE REVIEW

File: acquisition.py [hunk 1]
```diff
@@ -96,3 +108,144 @@ def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
...
+                        ]
+                    )
+                    .cpu()
+                    .detach()
+                    .numpy()
+                )
+        mean = np.mean(outputs, axis=1)
+        std = np.std(outputs, axis=1)
+        score = mean + self.config.acquisition.ucb.kappa * std
+        score = l2r(torch.Tensor(score))
```
Comment (manhbao-nguyen) [inline acquisition.py]: what is the role of l2r ?

File: acquisition.py [hunk 2]
```diff
@@ -96,3 +108,144 @@ def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
...
+
+    def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
+        super().get_reward_batch(inputs_af_base)
+
+        inputs_af = list(map(torch.tensor, inputs_af_base))
+        input_afs_lens = list(map(len, inputs_af_base))
+        inputs = pad_sequence(inputs_af, batch_first=True, padding_value=0.0)
+
+        self.load_best_proxy()
+        self.proxy.model.eval()
```
Comment (manhbao-nguyen) [inline acquisition.py]: To activate dropout, you'll have to do chose self.proxy.model.train(), no ?

File: acquisition.py [hunk 3]
```diff
@@ -96,3 +108,144 @@ def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
...
+                    )
+                    .cpu()
+                    .detach()
+                    .numpy()
+                )
+            elif self.config.proxy.model.lower() == "transformer":
+                outputs = (
+                    torch.hstack(
+                        [
+                            self.proxy.model(inputs, None)
```
Comment (manhbao-nguyen) [inline acquisition.py]: Do you think there would be any way to call the proxy in a very modular way ? like a general self.proxy.model(inputs, [optional parameters]) ? So that we don't have to distinguish in the acquisition function.

Comment (nikita-0209) [reply acquisition.py]: Yeah I've already created a PR for that :)
Can only modify these files once that is merged

File: acquisition.py (hunk 2 above)
Comment (nikita-0209) [reply acquisition.py]: Right sorry, my bad, I will correct that

File: acquisition.py (hunk 1 above)
Comment (nikita-0209) [reply acquisition.py]: According to my understanding, it is used for scaling the scores from the proxy. But I am unsure if this is actually required because `acq2proxy()` already does the scaling for us

File: acquisition.py [hunk 4]
```diff
@@ -86,7 +91,87 @@ def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
         inputs = pad_sequence(inputs_af, batch_first=True, padding_value=0.0)
 
         self.load_best_proxy()
-        self.proxy.model.eval()
+        self.proxy.model.train()
```
Comment (manhbao-nguyen) [inline acquisition.py:101]: Do we actually want to add dropout one-shot inference (proxy values without uncertainty ?)

File: acquisition.py [hunk 5]
```diff
@@ -86,7 +91,87 @@ def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
...
+            outputs = (
+                torch.hstack([self.proxy.model(inputs, input_afs_lens, None)])
+                .cpu()
+                .detach()
+                .numpy()
+            )
+
+        mean = np.mean(outputs, axis=1)
+        std = np.std(outputs, axis=1)
+        score = mean + self.config.acquisition.ucb.kappa * std
```
Comment (manhbao-nguyen) [inline acquisition.py:143]: Do we need to denormalize the mean ?  To me, there is no need, but we have to keep in mind (for further stats) that this is not the UCB score with the "oracle model" values, it is normalized.

File: acquisition.py [hunk 6]
```diff
@@ -86,7 +91,87 @@ def get_reward_batch(self, inputs_af_base):  # inputs_af = list of ...
...
+        super().__init__(config, proxy)
+
+    def load_best_proxy(self):
+        super().load_best_proxy()
+
+    def getMinF(self, inputs, input_len, mask):
+        # inputs = torch.Tensor(inputs).to(self.config.device)
+
+        if self.config.proxy.model.lower() == "mlp":
+            outputs = self.proxy.model(inputs).cpu().detach().numpy()
```
Comment (manhbao-nguyen) [inline acquisition.py]: Is there a way not be forced to distinguish different ways of calling the models : self.model(inputs, **kwargs) for all proxy models ?  That would remove the need to distinguish between different proxies for all acquisition functions

**Review by alexhernandezgarcia — APPROVED** [review]

File: acquisition.py (hunk 5 above)
Comment (nikita-0209) [reply acquisition.py:143]: I'm sorry not sure I understand your question: where is the mean being denormalised here?

File: acquisition.py (hunk 6 above)
Comment (nikita-0209) [reply acquisition.py]: Yeah yeah, you made a similar commenta few days ago. I forgot to update this branch when the PR for a uniform model was merged because I am not really working on this branch. But yes, as you suggest there is one way to call all proxy models (in the class specific get_reward_batch functions). I will update this branch with the correct code for this one function get_min_F

File: acquisition.py (hunk 4 above)
Comment (nikita-0209) [reply acquisition.py:101]: to be discussed I guess

File: acquisition.py (hunk 5 above)
Comment (manhbao-nguyen) [reply acquisition.py:143]: Like in the previous_code : actual_mean= stat_std* current_mean + stat_mean

Comment (nikita-0209) [reply acquisition.py:143]: Yeah I thought we discuss we should not denormalise it because we're ultimately gonna apply rhe reward/boltzmann function.

File: acquisition.py [hunk 7]
```diff
@@ -19,8 +20,14 @@ def __init__(self, config, proxy):
 
     def init_acquisition(self):
         # so far, only proxy acquisition function has been implemented, only add new acquisition class inheriting from AcquisitionFunctionBase to innovate
-        if self.config.acquisition.main == "proxy":
+        if self.config.acquisition.main.lower() == "proxy":
             self.acq = AcquisitionFunctionProxy(self.config, self.proxy)
+        elif self.config.acquisition.main == "oracle":
```
Comment (manhbao-nguyen) [inline acquisition.py]: self.config.acquisition.main.lower() == "oracle"

**Review by manhbao-nguyen — APPROVED** [review]
